src/main/scala/com/fulcrumgenomics/bam/pileup/ByCoordinatePileupBuilder.scala

```diff
@@ -0,0 +1,182 @@
+  override val mappedPairsOnly: Boolean                   = true,
+  override val includeDuplicates: Boolean                 = false,
+  override val includeSecondaryAlignments: Boolean        = false,
+  override val includeSupplementalAlignments: Boolean     = false,
+  override val excludeMapPositionOutsideFrInsert: Boolean = true,
+  val initialCacheSize: Int                               = DefaultInitialCacheSize,
+) extends PileupBuilder(
+  dict                              = dict,
+  minMapQ                           = minMapQ,
+  minBaseQ                          = minBaseQ,
+  mappedPairsOnly                   = mappedPairsOnly,
+  includeDuplicates                 = includeDuplicates,
+  includeSecondaryAlignments        = includeSecondaryAlignments,
+  includeSupplementalAlignments     = includeSupplementalAlignments,
+  excludeMapPositionOutsideFrInsert = excludeMapPositionOutsideFrInsert,
+) with Closeable { // TODO: One could extend AbstractIterator[Pileup[PileupEntry]] and scan the loci one-by-one as well!
+  import com.fulcrumgenomics.bam.pileup.ByCoordinatePileupBuilder.LocatablePileup
+
+  /** Whether this builder is able to pileup more records from the input iterator of SAM records or not. */
+  private var closed: Boolean = false
+
+  /** The last pileup we built over the input SAM records and is cached to save time if the locus is re-requested. */
+  private var lastPileup: Option[Pileup[PileupEntry]] = None
+
+  /** Advance this builder to the next requested locus and add all possibly overlapping records to the cache. */
+  @inline private def advanceTo(refIndex: Int, pos: Int): Unit = {
+    // Drop records up until the next record stands a chance of overlapping the requested locus. Then, take records up
+    // until the next record stands a chance of having a start greater than the requested locus plus one. All records in
+    // this query have a chance of overlapping the locus so we must then filter down to only those that have an end
+    // location greater than or equal to the requested locus. Finally, these records will be added to the cache so that
+    // they can be evaluated for pileup at the requested locus. We add one to <pos> for the case when a record starts
+    // with an insertion.
+    val maybeOverlapping = underlying
+      .dropWhile(rec => rec.refIndex < refIndex || (rec.refIndex == refIndex && rec.end < pos))
+      .takeWhile(rec => rec.refIndex == refIndex && rec.start <= pos + 1)
+      .filter(rec => rec.end >= pos)
+    cache.addAll(maybeOverlapping)
+  }
+
+  /** A genomic ordering for any locatable that utilizes the sequence dictionary corresponding to the input records. */
+  private lazy val byCoordinate: Ordering[Locatable] = GenomicOrdering(dict)
+
+  /** Records that we've accumulated that could overlap another coordinate-advancing call to <skipTo>. */
+  private lazy val cache: mutable.ArrayBuffer[SamRecord] = new mutable.ArrayBuffer[SamRecord](initialCacheSize)
```

Comment:
Profiling has revealed this to be the most efficient Scala collection for the cache.